deps/rabbitmq_stomp/src/rabbit_stomp_processor.erl

```diff
@@ -684,117 +690,383 @@ do_subscribe(Destination, DestHdr, Frame,
     Resource = #resource{virtual_host = VHost,
         kind = topic,
         name = rabbit_data_coercion:to_binary(Exchange)},
     Context = #{routing_key  => rabbit_data_coercion:to_binary(RoutingKey),
                 variable_map => #{<<"vhost">> => VHost, <<"username">> => Username}
     },
-    rabbit_access_control:check_topic_access(InternalUser, Resource, read, Context);
+    rabbit_access_control:check_topic_access(User, Resource, read, Context);
 check_subscription_access(_, _) ->
     authorized.
 
-maybe_clean_up_queue(Queue, #proc_state{connection = Connection}) ->
-    {ok, Channel} = amqp_connection:open_channel(Connection),
-    catch amqp_channel:call(Channel, #'queue.delete'{queue = Queue}),
-    catch amqp_channel:close(Channel),
+maybe_clean_up_queue(Queue, #proc_state{auth_login = Username}) ->
+    catch delete_queue(Queue, Username),
     ok.
 
 do_send(Destination, _DestHdr,
         Frame = #stomp_frame{body_iolist = BodyFragments},
-        State = #proc_state{channel = Channel,
-                            route_state = RouteState,
-                            default_topic_exchange = DfltTopicEx}) ->
-    case ensure_endpoint(dest, Destination, Frame, Channel, RouteState) of
-
-        {ok, _Q, RouteState1} ->
-
+        State0 = #proc_state{default_topic_exchange = DfltTopicEx,
+                             delivery_flow = Flow,
+                             user = User,
+                             authz_context = AuthzCtx,
+                             vhost = VHost}) ->
+    case ensure_endpoint(dest, Destination, Frame, State0) of
+
+        {ok, _Q, State} ->
             {Frame1, State1} =
-                ensure_reply_to(Frame, State#proc_state{route_state = RouteState1}),
+                ensure_reply_to(Frame, State),
 
             Props = rabbit_stomp_util:message_properties(Frame1),
 
-            {Exchange, RoutingKey} = parse_routing(Destination, DfltTopicEx),
-
-            Method = #'basic.publish'{
-              exchange = list_to_binary(Exchange),
-              routing_key = list_to_binary(RoutingKey),
-              mandatory = false,
-              immediate = false},
-
-            case transactional(Frame1) of
-                {yes, Transaction} ->
-                    extend_transaction(
-                      Transaction,
-                      fun(StateN) ->
-                              maybe_record_receipt(Frame1, StateN)
-                      end,
-                      {Method, Props, BodyFragments},
-                      State1);
-                no ->
-                    ok(send_method(Method, Props, BodyFragments,
-                                   maybe_record_receipt(Frame1, State1)))
-            end;
+            {ExchangeNameList, RoutingKeyList} = parse_routing(Destination, DfltTopicEx),
+            %% io:format("Parse_routing: ~p~n", [{ExchangeNameList, RoutingKeyList}]),
+            RoutingKey = list_to_binary(RoutingKeyList),
```

Comment:
The STOMP plugin operates too much on lists instead of binaries which is inefficient. At some point we should improve this. However, this should be a follow up PR.